common/lib/xmodule/xmodule/seq_module.py

```diff
@@ -279,6 +279,74 @@ def verify_current_content_visibility(cls, date, hide_after_date):
+        This was an issue because audit learners could start a timed exam
+        and then be unable to complete it because the graded content would be gated.
+        Even if they later upgraded, they could still be unable to complete the exam
+        because the timer could have expired.
+
+        Solution:
+        Gate the entire sequence when we think the above problem can occur.
+
+        If:
+        1. This sequence is a timed exam
+        2. And this sequence contains problems which this user cannot load due to content type gating
+        Then:
+        We will gate access to the entire sequence.
+        Otherwise, learners would have the ability to start their timer for an exam,
+        but then not have the ability to complete it.
+
+        We are displaying the gating fragment within the sequence, as is done for gating for prereqs,
+        rather than content type gating the entire sequence because that would remove the next/previous navigation.
+
+        This functionality still needs to be replicated in the frontend-app-learning courseware MFE
+        The ticket to track this is https://openedx.atlassian.net/browse/REV-1220
+        Note that this will break compatability with using sequences outside of edx-platform
+        but we are ok with this for now
+        """
+        # importing here to avoid a circular import
+        from openedx.features.content_type_gating.models import ContentTypeGatingConfig
+
+        if not self.is_time_limited:
+            return
+
+        try:
+            user = User.objects.get(id=self.runtime.user_id)
+            if not ContentTypeGatingConfig.enabled_for_enrollment(user=user, course_key=self.runtime.course_id):
+                return
+
+            for vertical in self.get_children():
+                for block in vertical.get_children():
+                    problem_eligible_for_content_gating = (getattr(block, 'graded', False) and
+                                                           block.has_score and
+                                                           getattr(block, 'weight', 0) != 0)
+                    if problem_eligible_for_content_gating:
+                        access = self._has_access_check(user, 'load', block, self.course_id)
+                        # If any block has been gated by content type gating inside the sequence
+                        # and the sequence is a timed exam, then gate the entire sequence.
+                        # In order to avoid scope creep, we are not handling other potential causes
+                        # of access failures as part of this work.
+                        if not access and access.error_code == 'incorrect_user_group':
+                            self.gated_sequence_fragment = access.user_fragment
+                            break
+                        self.gated_sequence_fragment = None  # Don't gate other cases
```

Comment:
Does this also not start the clock for the audit user when they get here/the course is not exempt?

Reply:
can you clarify the case you're referring to?

Reply:
My understanding that part of the issue was that when an audit learner goes to the gated content page, that they would start the timer on the timed exam. However, the timer on the timed exams starts on the page, no matter if they have access or not. By the time they finished purchasing access to the course, they would have lost a bunch of time for the timed exam. Does this code also prevent the timer from starting on the timed exam, if the user does not have access?

Reply:
if the gated_sequence_fragment is None then the content is not gated

Reply:
Per conversation this morning, it sounded like the timer does -not- start until the user manually starts it. The user won't have access to start the timer if the content is gated, so the issue with the timer starting for a user who doesn't have access to the timed exam should not be an issue.